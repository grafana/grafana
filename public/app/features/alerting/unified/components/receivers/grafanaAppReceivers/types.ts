import { SupportedPlugin } from '../../../types/pluginBridges';

export const GRAFANA_APP_RECEIVERS_SOURCE_IMAGE: Record<SupportedPlugin, string> = {
  [SupportedPlugin.OnCall]: 'public/img/alerting/oncall_logo.svg',

  [SupportedPlugin.Incident]: '',
  [SupportedPlugin.MachineLearning]: '',
  [SupportedPlugin.Labels]: '',
  [SupportedPlugin.Slo]: '',
};
