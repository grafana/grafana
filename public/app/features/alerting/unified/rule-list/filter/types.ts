import { PromAlertingRuleState, PromRuleType } from 'app/types/unified-alerting-dto';

import type { RuleHealth } from '../../search/rulesSearchParser';

export type AdvancedFilters = {
  namespace?: string | null;
  groupName?: string | null;
  ruleName?: string;
  ruleType?: PromRuleType | '*';
  ruleState: PromAlertingRuleState | '*';
  dataSourceNames: string[];
  labels: string[];
  ruleHealth?: RuleHealth | '*';
  dashboardUid?: string;
  plugins?: 'show' | 'hide';
  contactPoint?: string | null;
};
