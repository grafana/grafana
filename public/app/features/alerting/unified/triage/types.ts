export type Domain = [Date, Date];
export type Filter = [key: string, operator: '=' | '=!', value: string];

export type WorkbenchRow = GenericGroupedRow | AlertRuleRow;

export type TimelineEntry = [timestamp: number, state: 'firing' | 'pending'];

export interface AlertRuleRow {
  type: 'alertRule';
  metadata: {
    title: string;
    folder: string;
    ruleUID: string;
  };
}

export interface GenericGroupedRow {
  type: 'group';
  metadata: {
    label: string;
    value: string;
  };
  rows: WorkbenchRow[];
}
