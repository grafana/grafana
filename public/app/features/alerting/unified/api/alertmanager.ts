import { lastValueFrom } from 'rxjs';

import { isObject } from '@grafana/data';
import { getBackendSrv, isFetchError } from '@grafana/runtime';
import {
  AlertManagerCortexConfig,
  AlertmanagerGroup,
  AlertmanagerStatus,
  Receiver,
  TestReceiversAlert,
  TestReceiversPayload,
  TestReceiversResult,
} from 'app/plugins/datasource/alertmanager/types';

import { GRAFANA_RULES_SOURCE_NAME, getDatasourceAPIUid } from '../utils/datasource';

// "grafana" for grafana-managed, otherwise a datasource name
export async function fetchAlertManagerConfig(alertManagerSourceName: string): Promise<AlertManagerCortexConfig> {
  try {
    const result = await lastValueFrom(
      getBackendSrv().fetch<AlertManagerCortexConfig>({
        url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/alerts`,
        showErrorAlert: false,
        showSuccessAlert: false,
      })
    );
    return {
      template_files: result.data.template_files ?? {},
      template_file_provenances: result.data.template_file_provenances ?? {},
      alertmanager_config: result.data.alertmanager_config ?? {},
      last_applied: result.data.last_applied,
      id: result.data.id,
    };
  } catch (e) {
    // if no config has been uploaded to grafana, it returns error instead of latest config
    if (
      alertManagerSourceName === GRAFANA_RULES_SOURCE_NAME &&
      isFetchError(e) &&
      e.data?.message?.includes('could not find an Alertmanager configuration')
    ) {
      return {
        template_files: {},
        alertmanager_config: {},
      };
    }
    throw e;
  }
}

export async function updateAlertManagerConfig(
  alertManagerSourceName: string,
  config: AlertManagerCortexConfig
): Promise<void> {
  await lastValueFrom(
    getBackendSrv().fetch({
      method: 'POST',
      url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/alerts`,
      data: config,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );
}

export async function deleteAlertManagerConfig(alertManagerSourceName: string): Promise<void> {
  await lastValueFrom(
    getBackendSrv().fetch({
      method: 'DELETE',
      url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/alerts`,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );
}

export async function fetchAlertGroups(alertmanagerSourceName: string): Promise<AlertmanagerGroup[]> {
  const result = await lastValueFrom(
    getBackendSrv().fetch<AlertmanagerGroup[]>({
      url: `/api/alertmanager/${getDatasourceAPIUid(alertmanagerSourceName)}/api/v2/alerts/groups`,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );

  return result.data;
}

export async function fetchStatus(alertManagerSourceName: string): Promise<AlertmanagerStatus> {
  const result = await lastValueFrom(
    getBackendSrv().fetch<AlertmanagerStatus>({
      url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/api/v2/status`,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );

  return result.data;
}

export async function testReceivers(
  alertManagerSourceName: string,
  receivers: Receiver[],
  alert?: TestReceiversAlert
): Promise<void> {
  const data: TestReceiversPayload = {
    receivers,
    alert,
  };
  try {
    const result = await lastValueFrom(
      getBackendSrv().fetch<TestReceiversResult>({
        method: 'POST',
        data,
        url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/receivers/test`,
        showErrorAlert: false,
        showSuccessAlert: false,
      })
    );

    if (receiversResponseContainsErrors(result.data)) {
      throw new Error(getReceiverResultError(result.data));
    }
  } catch (error) {
    if (isFetchError(error) && isTestReceiversResult(error.data) && receiversResponseContainsErrors(error.data)) {
      throw new Error(getReceiverResultError(error.data));
    }

    throw error;
  }
}

function receiversResponseContainsErrors(result: TestReceiversResult) {
  return result.receivers.some((receiver) =>
    receiver.grafana_managed_receiver_configs.some((config) => config.status === 'failed')
  );
}

function isTestReceiversResult(data: unknown): data is TestReceiversResult {
  if (isObject(data) && 'receivers' in data && Array.isArray(data.receivers)) {
    return data.receivers.every(isSingleTestRecieverResult);
  }

  return false;
}

function isSingleTestRecieverResult(receiver: unknown): receiver is TestReceiversResult {
  return (
    isObject(receiver) &&
    'name' in receiver &&
    typeof receiver.name === 'string' &&
    'grafana_managed_receiver_configs' in receiver &&
    Array.isArray(receiver.grafana_managed_receiver_configs)
  );
}

function getReceiverResultError(receiversResult: TestReceiversResult) {
  return receiversResult.receivers
    .flatMap((receiver) =>
      receiver.grafana_managed_receiver_configs
        .filter((receiver) => receiver.status === 'failed')
        .map((receiver) => receiver.error ?? 'Unknown error.')
    )
    .join('; ');
}
