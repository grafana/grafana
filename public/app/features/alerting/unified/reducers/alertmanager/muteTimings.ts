import { createAction, createReducer } from '@reduxjs/toolkit';
import { remove } from 'lodash';

import { AlertManagerCortexConfig, MuteTimeInterval } from 'app/plugins/datasource/alertmanager/types';

import { removeTimeIntervalFromRoute, renameTimeInterval } from '../../utils/alertmanager';

export const addMuteTimingAction = createAction<{ interval: MuteTimeInterval }>('muteTiming/add');
export const updateMuteTimingAction = createAction<{
  interval: MuteTimeInterval;
  originalName: string;
}>('muteTiming/update');
export const deleteMuteTimingAction = createAction<{ name: string }>('muteTiming/delete');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to mute timings and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const muteTimingsReducer = createReducer(initialState, (builder) => {
  builder
    // add a mute timing to the alertmanager configuration
    .addCase(addMuteTimingAction, (draft, { payload }) => {
      const { interval } = payload;
      draft.alertmanager_config.time_intervals = (draft.alertmanager_config.time_intervals ?? []).concat(interval);
    })
    // add a mute timing to the alertmanager configuration
    // make sure we update the mute timing in either the deprecated or the new time intervals property
    .addCase(updateMuteTimingAction, (draft, { payload }) => {
      const { interval, originalName } = payload;
      const nameHasChanged = interval.name !== originalName;

      const timeIntervals = draft.alertmanager_config.time_intervals ?? [];
      const muteTimeIntervals = draft.alertmanager_config.mute_time_intervals ?? [];

      const existingIntervalIndex = timeIntervals.findIndex(({ name }) => name === originalName);
      if (existingIntervalIndex !== -1) {
        timeIntervals[existingIntervalIndex] = interval;
      }

      const existingMuteIntervalIndex = muteTimeIntervals.findIndex(({ name }) => name === originalName);
      if (existingMuteIntervalIndex !== -1) {
        muteTimeIntervals[existingMuteIntervalIndex] = interval;
      }

      if (nameHasChanged && draft.alertmanager_config.route) {
        draft.alertmanager_config.route = renameTimeInterval(
          interval.name,
          originalName,
          draft.alertmanager_config.route
        );
      }
    })
    // delete a mute timing from the alertmanager configuration, since the configuration might be using the "deprecated" mute_time_intervals
    // let's also check there
    .addCase(deleteMuteTimingAction, (draft, { payload }) => {
      const { name } = payload;
      const { alertmanager_config } = draft;
      const { time_intervals = [], mute_time_intervals = [] } = alertmanager_config;

      // remove the mute timings from the legacy and new time intervals definition
      remove(time_intervals, (interval) => interval.name === name);
      remove(mute_time_intervals, (interval) => interval.name === name);

      // remove the mute timing from all routes
      alertmanager_config.route = removeTimeIntervalFromRoute(name, alertmanager_config.route ?? {});
    });
});
