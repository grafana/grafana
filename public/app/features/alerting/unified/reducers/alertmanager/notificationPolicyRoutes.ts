import { createAction, createReducer } from '@reduxjs/toolkit';

import { AlertManagerCortexConfig } from 'app/plugins/datasource/alertmanager/types';

import { FormAmRoute } from '../../types/amroutes';
import { addUniqueIdentifierToRoute } from '../../utils/amroutes';
import {
  InsertPosition,
  addRouteToReferenceRoute,
  cleanRouteIDs,
  mergePartialAmRouteWithRouteTree,
  omitRouteFromRouteTree,
} from '../../utils/routeTree';

export const updateRouteAction = createAction<{ update: Partial<FormAmRoute>; alertmanager: string }>('routes/update');
export const deleteRouteAction = createAction<{ id: string }>('routes/delete');
export const addRouteAction = createAction<{
  alertmanager: string;
  partialRoute: Partial<FormAmRoute>;
  referenceRouteIdentifier: string;
  insertPosition: InsertPosition;
}>('routes/add');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to routes and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const routesReducer = createReducer(initialState, (builder) => {
  builder
    // update an existing route
    .addCase(updateRouteAction, (draft, { payload }) => {
      const { update, alertmanager } = payload;
      const { alertmanager_config } = draft;

      if (!alertmanager_config.route) {
        return;
      }

      const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(alertmanager_config.route);
      const newRouteTree = mergePartialAmRouteWithRouteTree(alertmanager, update, rootRouteWithIdentifiers);
      alertmanager_config.route = cleanRouteIDs(newRouteTree);
    })
    // delete a route
    .addCase(deleteRouteAction, (draft, { payload }) => {
      const { id } = payload;
      const { alertmanager_config } = draft;

      // if we don't even have a root route, we bail
      if (!alertmanager_config.route) {
        return;
      }

      const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(alertmanager_config.route);
      const updatedPolicyTree = omitRouteFromRouteTree(id, rootRouteWithIdentifiers);
      draft.alertmanager_config.route = cleanRouteIDs(updatedPolicyTree);
    })
    // add a new route to given position
    .addCase(addRouteAction, (draft, { payload }) => {
      const { partialRoute, referenceRouteIdentifier, insertPosition, alertmanager } = payload;
      const { alertmanager_config } = draft;

      if (!alertmanager_config.route) {
        return;
      }

      const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(alertmanager_config.route);
      const updatedPolicyTree = addRouteToReferenceRoute(
        alertmanager,
        partialRoute,
        referenceRouteIdentifier,
        rootRouteWithIdentifiers,
        insertPosition
      );

      draft.alertmanager_config.route = cleanRouteIDs(updatedPolicyTree);
    });
});
