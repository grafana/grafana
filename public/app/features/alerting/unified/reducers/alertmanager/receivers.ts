import { createAction, createReducer } from '@reduxjs/toolkit';
import { remove } from 'lodash';

import { AlertManagerCortexConfig, Receiver } from 'app/plugins/datasource/alertmanager/types';

import { renameReceiverInRoute } from '../../utils/notification-policies';

export const addReceiverAction = createAction<Receiver>('receiver/add');
export const updateReceiverAction = createAction<{ name: string; receiver: Receiver }>('receiver/update');
export const deleteReceiverAction = createAction<string>('receiver/delete');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to receiver (Contact points) and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const receiversReducer = createReducer(initialState, (builder) => {
  builder
    // add a new receiver
    .addCase(addReceiverAction, (draft, { payload: newReceiver }) => {
      // ensure the receivers are always an array
      const currentReceivers = (draft.alertmanager_config.receivers = draft.alertmanager_config.receivers ?? []);

      // check if the name doesn't already exist
      const nameExists = currentReceivers.some((receiver) => receiver.name === newReceiver.name);
      if (nameExists) {
        throw new Error(`Duplicate receiver name ${newReceiver.name}`);
      }

      // add the receiver
      currentReceivers.push(newReceiver);
    })
    // upate an existing receiver
    .addCase(updateReceiverAction, (draft, { payload }) => {
      const { name, receiver } = payload;
      const renaming = name !== receiver.name;

      const receivers = draft.alertmanager_config.receivers ?? [];

      const targetIndex = receivers.findIndex((receiver) => receiver.name === name);
      const targetExists = targetIndex > -1;

      // check if the receiver we want to update exists
      if (!targetExists) {
        throw new Error(`Expected receiver ${name} to exist, but did not find it in the config`);
      }

      // check if the new name doesn't already exist
      if (renaming) {
        const nameExists = receivers.some((oldReceiver) => oldReceiver.name === receiver.name);
        if (nameExists) {
          throw new Error(`Duplicate receiver name ${receiver.name}`);
        }
      }

      // overwrite the receiver with the new one
      receivers[targetIndex] = receiver;

      // check if we need to update routes if the contact point was renamed
      const routeTree = draft.alertmanager_config.route;

      if (routeTree && renaming) {
        draft.alertmanager_config.route = renameReceiverInRoute(routeTree, name, receiver.name);
      }
    })
    // delete a receiver from the alertmanager configuration
    .addCase(deleteReceiverAction, (draft, { payload: name }) => {
      remove(draft.alertmanager_config.receivers ?? [], (receiver) => receiver.name === name);
    });
});
