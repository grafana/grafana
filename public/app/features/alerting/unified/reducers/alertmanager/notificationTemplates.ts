import { createAction, createReducer } from '@reduxjs/toolkit';
import { remove, toArray, unset } from 'lodash';

import { AlertManagerCortexConfig } from 'app/plugins/datasource/alertmanager/types';

import { TemplateFormValues } from '../../components/receivers/TemplateForm';
import { ensureDefine } from '../../utils/templates';

export const addNotificationTemplateAction = createAction<{ template: TemplateFormValues }>('notificationTemplate/add');
export const updateNotificationTemplateAction = createAction<{
  name: string;
  template: TemplateFormValues;
}>('notificationTemplate/update');
export const deleteNotificationTemplateAction = createAction<{ name: string }>('notificationTemplate/delete');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to notification templates and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const notificationTemplatesReducer = createReducer(initialState, (builder) => {
  builder
    .addCase(addNotificationTemplateAction, (draft, { payload }) => {
      const { alertmanager_config = {}, template_files = {} } = draft;
      const { template } = payload;

      const targetTemplateExists = template_files[template.title] !== undefined;
      if (targetTemplateExists) {
        throw new Error('target template already exists');
      }

      // wrap content in "define" if it's not already wrapped, in case user did not do it/
      // it's not obvious that this is needed for template to work
      const content = ensureDefine(template.title, template.content);

      // add the template to the list of template files
      template_files[template.title] = content;

      // add the template to the alertmanager_config
      alertmanager_config.templates = toArray(alertmanager_config.templates).concat(template.title);
    })
    .addCase(updateNotificationTemplateAction, (draft, { payload }) => {
      const { alertmanager_config = {}, template_files = {} } = draft;
      const { name, template } = payload;
      const renaming = name !== template.title;

      const targetExists = template_files[name] !== undefined;
      if (!targetExists) {
        throw new Error(`Expected notification template ${name} to exist, but did not find it in the config`);
      }

      // wrap content in "define" if it's not already wrapped, in case user did not do it/
      // it's not obvious that this is needed for template to work
      const content = ensureDefine(template.title, template.content);

      if (renaming) {
        const oldName = name;
        const newName = template.title;

        const targetExists = template_files[newName] !== undefined;
        if (targetExists) {
          throw new Error(`Duplicate template name ${newName}`);
        }

        unset(template_files, oldName);
        remove(alertmanager_config.templates ?? [], (templateName) => templateName === oldName);
        alertmanager_config.templates = toArray(alertmanager_config.templates).concat(template.title);
      }

      template_files[template.title] = content;
    })
    .addCase(deleteNotificationTemplateAction, (draft, { payload }) => {
      const { name } = payload;
      const { alertmanager_config = {}, template_files = {} } = draft;

      unset(template_files, name);
      remove(alertmanager_config.templates ?? [], (templateName) => templateName === name);
    });
});
