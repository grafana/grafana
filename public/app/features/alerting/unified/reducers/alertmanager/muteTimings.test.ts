import { AlertManagerCortexConfig, MuteTimeInterval } from 'app/plugins/datasource/alertmanager/types';

import { addMuteTimingAction, deleteMuteTimingAction, muteTimingsReducer, updateMuteTimingAction } from './muteTimings';

describe('mute timings', () => {
  const initialConfig: AlertManagerCortexConfig = {
    alertmanager_config: {
      time_intervals: [mockTimeInterval({ name: 'default time interval' })],
      mute_time_intervals: [mockTimeInterval({ name: 'default legacy time interval' })],
      route: {
        routes: [
          {
            mute_time_intervals: ['default time interval'],
          },
          {
            mute_time_intervals: ['default legacy time interval'],
          },
        ],
      },
    },
    template_files: {},
  };

  it('should be able to add a new mute timing', () => {
    const newMuteTiming = mockTimeInterval({ name: 'new mute time interval' });
    const action = addMuteTimingAction({ interval: newMuteTiming });

    expect(muteTimingsReducer(initialConfig, action)).toMatchSnapshot();
  });

  it('should be able to remove a mute timing', () => {
    const legacyMuteTimingName = initialConfig.alertmanager_config.mute_time_intervals![0].name;
    const deleteFromLegacyKey = deleteMuteTimingAction({ name: legacyMuteTimingName });
    expect(muteTimingsReducer(initialConfig, deleteFromLegacyKey)).toMatchSnapshot();

    const muteTimingName = initialConfig.alertmanager_config.time_intervals![0].name;
    const deleteMuteTiming = deleteMuteTimingAction({ name: muteTimingName });
    expect(muteTimingsReducer(initialConfig, deleteMuteTiming)).toMatchSnapshot();
  });

  it('should be able to update a time interval', () => {
    const newMuteTiming = mockTimeInterval({ name: 'new mute time interval' });

    const legacyMuteTimingName = initialConfig.alertmanager_config.mute_time_intervals![0].name;
    const updateLegacyMuteTiming = updateMuteTimingAction({
      originalName: legacyMuteTimingName,
      interval: newMuteTiming,
    });
    expect(muteTimingsReducer(initialConfig, updateLegacyMuteTiming)).toMatchSnapshot();

    const muteTimingName = initialConfig.alertmanager_config.time_intervals![0].name;
    const updateMuteTiming = updateMuteTimingAction({ originalName: muteTimingName, interval: newMuteTiming });
    expect(muteTimingsReducer(initialConfig, updateMuteTiming)).toMatchSnapshot();
  });
});

function mockTimeInterval(overrides: Partial<MuteTimeInterval> = {}): MuteTimeInterval {
  return {
    name: 'mock time interval',
    time_intervals: [
      {
        times: [
          {
            start_time: '12:00',
            end_time: '24:00',
          },
        ],
        days_of_month: ['15'],
        months: ['august:december', 'march'],
      },
    ],
    ...overrides,
  };
}
