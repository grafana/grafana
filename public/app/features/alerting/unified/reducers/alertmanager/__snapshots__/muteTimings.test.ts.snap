// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mute timings should be able to add a new mute timing 1`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [
      {
        "name": "default legacy time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
    "route": {
      "routes": [
        {
          "mute_time_intervals": [
            "default time interval",
          ],
        },
        {
          "mute_time_intervals": [
            "default legacy time interval",
          ],
        },
      ],
    },
    "time_intervals": [
      {
        "name": "default time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
      {
        "name": "new mute time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
  },
  "template_files": {},
}
`;

exports[`mute timings should be able to remove a mute timing 1`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [],
    "route": {
      "active_time_intervals": [],
      "mute_time_intervals": [],
      "routes": [
        {
          "active_time_intervals": [],
          "mute_time_intervals": [
            "default time interval",
          ],
          "routes": undefined,
        },
        {
          "active_time_intervals": [],
          "mute_time_intervals": [],
          "routes": undefined,
        },
      ],
    },
    "time_intervals": [
      {
        "name": "default time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
  },
  "template_files": {},
}
`;

exports[`mute timings should be able to remove a mute timing 2`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [
      {
        "name": "default legacy time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
    "route": {
      "active_time_intervals": [],
      "mute_time_intervals": [],
      "routes": [
        {
          "active_time_intervals": [],
          "mute_time_intervals": [],
          "routes": undefined,
        },
        {
          "active_time_intervals": [],
          "mute_time_intervals": [
            "default legacy time interval",
          ],
          "routes": undefined,
        },
      ],
    },
    "time_intervals": [],
  },
  "template_files": {},
}
`;

exports[`mute timings should be able to update a time interval 1`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [
      {
        "name": "new mute time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
    "route": {
      "active_time_intervals": undefined,
      "mute_time_intervals": undefined,
      "routes": [
        {
          "active_time_intervals": undefined,
          "mute_time_intervals": [
            "default time interval",
          ],
          "routes": undefined,
        },
        {
          "active_time_intervals": undefined,
          "mute_time_intervals": [
            "new mute time interval",
          ],
          "routes": undefined,
        },
      ],
    },
    "time_intervals": [
      {
        "name": "default time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
  },
  "template_files": {},
}
`;

exports[`mute timings should be able to update a time interval 2`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [
      {
        "name": "default legacy time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
    "route": {
      "active_time_intervals": undefined,
      "mute_time_intervals": undefined,
      "routes": [
        {
          "active_time_intervals": undefined,
          "mute_time_intervals": [
            "new mute time interval",
          ],
          "routes": undefined,
        },
        {
          "active_time_intervals": undefined,
          "mute_time_intervals": [
            "default legacy time interval",
          ],
          "routes": undefined,
        },
      ],
    },
    "time_intervals": [
      {
        "name": "new mute time interval",
        "time_intervals": [
          {
            "days_of_month": [
              "15",
            ],
            "months": [
              "august:december",
              "march",
            ],
            "times": [
              {
                "end_time": "24:00",
                "start_time": "12:00",
              },
            ],
          },
        ],
      },
    ],
  },
  "template_files": {},
}
`;
